### PR TITLE
Feature/add tests for lib directory

### PR DESCRIPTION
**Pull request by AndrzejSala**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Add tests for /lib except upload.ts (the upload class will be soon refactored)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Other information**:
Add some fixme annotations to consider later

### CODE REVIEW

File: src/lib/api/preview.spec.ts [hunk 1]
```diff
@@ -2,91 +2,155 @@
...
-  it('should throw exception when handle is not provided', () => {
-    assert.throws(() => preview(session));
-  });
-
-  it('should open window when no id specified', () => {
-    preview(session, filelink);
-    assert.ok(windowOpenSpy.calledWith(`${session.urls.cdnUrl}/preview/${filelink}`, filelink));
-  });
+    it('should open a proper url with preview of a file with src handle', () => {
+      const defaultSession = {
```
Comment (pcholuj) [inline src/lib/api/preview.spec.ts]: Define default session before all tests, you dont have to repeat it in all tests

File: src/lib/api/security.ts [hunk 2]
```diff
@@ -64,6 +64,7 @@ export const getSecurity = (policyOptions: SecurityOptions, appSecret: string):
 
   checkOptions('Policy options', allowed, policyOptions);
 
+  // FIXME: https://nodejs.org/en/docs/guides/buffer-constructor-deprecation/
```
Comment (pcholuj) [inline src/lib/api/security.ts]: You can change it to buffer.from

File: src/lib/api/upload/file_utils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
... 251, 175, 184, 74, 158, 129, 246, 11, 46, 254, 191, 251, 246, 219, 127, 116, 144, 183, 229, 22, 97, 3, 101, 103, 156, 197, 16, 92, 119, 249, 213, 127, 108, 128, 144, 129, 32, 179, 24, 24, 138, 118, 255, 103, 168, 63, 248, 159, 161, 251, 248, 127, 190, 169, 103, 192, 18, 232, 174, 59, 250, 224, 3, 78, 47, 131, 244, 128, 244, 130, 205, 0, 153, 5, 38, 144, 113, 233, 30, 12, 151, 49, 180, 29, 65, 197, 141, 135, 255, 51, 212, 28, 0, 171, 197, 208, 15, 103, 244, 158, 248, 191, 255, 238, 59, 48, 70, 7, 48, 113, 144, 26, 12, 3, 112, 25, 40, 52, 237, 204, 127, 66, 0, 164, 102, 224, 12, 100, 168, 62, 0, 214, 128, 205, 96, 152, 56, 72, 13, 241, 6, 34, 97, 140, 72, 33, 100, 200, 168, 129, 247, 169, 104, 224, 125, 144, 129, 1, 84, 52, 48, 128, 1, 12, 32, 134, 238, 135, 97, 44, 6, 238, 199, 131, 223, 67, 125, 9, 54, 12, 0, 87, 126, 31, 109, 180, 228, 26, 37, 0, 0, 0, 25, 116, 69, 88, 116, 83, 111, 102, 116, 119, 97, 114, 101, 0, 65, 100, 111, 98, 101, 32, 73, 109, 97, 103, 101, 82, 101, 97, 100, 121, 113, 201, 101, 60, 0, 0, 0, 18, 116, 69, 88, 116, 69, 88, 73, 70, 58, 79, 114, 105, 101, 110, 116, 97, 116, 105, 111, 110, 0, 49, 132, 88, 236, 239, 0, 0, 0, 0, 73, 69, 78, 68, 174, 66, 96, 130]));
```
Comment (pcholuj) [inline src/lib/api/upload/file_utils.spec.ts]: For that u can define some minimal buffer before tests, and use it as return value of readFile and expect value in test case, You are checking only if buffer is equal so it dont have to be jpeg file. ie just Buffer.from('teststring')

File: src/lib/api/upload/file_utils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+      } else if (path === './../logo_png.1.png') {
+        cb({ 'errno': -2, 'code': 'ENOENT', 'syscall': 'open', 'path': '/Users/andrzej/workspace/filestack-js/src/lib/api/upload/file_utils_test_files/logo_png.1.png' }, null);
```
Comment (pcholuj) [inline src/lib/api/upload/file_utils.spec.ts]: do not hardcode that kind of path, u used ure username etc

File: src/lib/api/security.ts (hunk 2 above)
Comment (AndrzejSala) [reply src/lib/api/security.ts]: Ok

File: src/lib/api/upload/file_utils.spec.ts (hunk 3 above)
Comment (AndrzejSala) [reply src/lib/api/upload/file_utils.spec.ts]: Ok

File: src/lib/api/upload/file_utils.spec.ts (hunk 4 above)
Comment (AndrzejSala) [reply src/lib/api/upload/file_utils.spec.ts]: ok

File: src/lib/api/preview.spec.ts (hunk 1 above)
Comment (AndrzejSala) [reply src/lib/api/preview.spec.ts]: ok

File: src/lib/utils/index.spec.ts [hunk 5]
```diff
@@ -17,7 +14,122 @@
...
+      const handle = 'src://test123/flug_8-trans_atlantik-300dpi.jpg';
+      const result = resolveCdnUrl(session, handle);
+      expect(result).toEqual('https://cdn.fs.enplug.com/DevZpGbhtSzZL5bOtYw7Lz');
+    });
+    it('should properly resolve cdn url with hashed handle', () => {
+      const handle = 'https://static1.squarespace.com/static/544eb3cce4b0ef091773611f/t/59ba7ce1bd10f00dcdc80a5f/1505394087367/DSC_0527.jpg';
+      const result = resolveCdnUrl(session, handle);
+      expect(result).toEqual('https://cdn.fs.enplug.com/DevZpGbhtSzZL5bOtYw7Lz');
+    });
+    it('shoudl throw error when using src handle or url without apikey', () => {
```
Comment (maryfs) [inline src/lib/utils/index.spec.ts]: 'should throw an error ..'

File: src/lib/api/cloud.spec.ts [hunk 6]
```diff
@@ -15,6 +15,88 @@
...
+      };
+    }),
+  };
+});
+
+describe('api:cloud', () => {
+  describe('CloudClient', () => {
+    const createCloudClient = (clientOptions = {}) => {
+      const session = {
+        'apikey': 'AJrB9kEilS1SQeHIDf3wtz',
```
Comment (maryfs) [inline src/lib/api/cloud.spec.ts]: change apikey

File: src/lib/utils/index.spec.ts (hunk 5 above)
Comment (AndrzejSala) [reply src/lib/utils/index.spec.ts]: Fixed

File: src/lib/api/cloud.spec.ts (hunk 6 above)
Comment (AndrzejSala) [reply src/lib/api/cloud.spec.ts]: Fixed

File: src/lib/client.spec.ts [hunk 7]
```diff
@@ -15,6 +15,264 @@
...
+  it('should pass cname to the session', () => {
+    const options = {
+      cname: 'testCname.com',
+    };
+    const client = new Client(defaultApikey, options);
+    expect(client.session.cname).toBe(options.cname);
+  });
+  it('should be able to logout for provied cloud', (done) => {
+    const client = new Client(defaultApikey);
+    const cloudName = 'dropboxx';
```
Comment (maryfs) [inline src/lib/client.spec.ts]: typo dropbox

File: src/lib/client.spec.ts [hunk 8]
```diff
@@ -15,6 +15,264 @@
...
+    },
+  };
+});
+
+describe('client', () => {
+  const defaultApikey = 'CmrB9kEilS1SQeHIDf3wtz';
+  const defaultHandle = '5aYkEQJSQCmYShsoCnZN';
+  const defaultSecurity = {
+    policy: 'examplePoolicy',
+    signature: 'blablabla',
```
Comment (maryfs) [inline src/lib/client.spec.ts]: exampleSignature :)

File: src/lib/client.spec.ts [hunk 9]
```diff
@@ -15,6 +15,264 @@
...
+  it('should be able to open preview', () => {
+    const client = new Client(defaultApikey);
+    const previewOptions = {
+      id: 'testElement',
+    };
+    client.preview(defaultHandle, previewOptions);
+    expect(mockPreview).toBeCalled();
+    expect(mockPreview).toBeCalledWith(expect.any(Object), defaultHandle, previewOptions);
+  });
+  it('should be able to remove handle', (done) => {
```
Comment (maryfs) [inline src/lib/client.spec.ts]: please add another test for remove (should throw an error during remove if security is not given)

File: src/lib/api/security.spec.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+import { getSecurity } from './security';
+
+describe('api:security', () => {
+  describe('getSecurity', () => {
+    it('should create a proper object with policy & signature', () => {
+      const policy = {
+        expiry: 1523595600,
+        call: ['read', 'convert'],
```
Comment (maryfs) [inline src/lib/api/security.spec.ts]: add all calls:
["pick","read","stat","write","writeUrl","store","convert","remove","exif","runWorkflow"]

File: src/lib/client.spec.ts (hunk 7 above)
Comment (AndrzejSala) [reply src/lib/client.spec.ts]: Fixed

File: src/lib/client.spec.ts (hunk 8 above)
Comment (AndrzejSala) [reply src/lib/client.spec.ts]: Fixed

File: src/lib/client.spec.ts (hunk 9 above)
Comment (AndrzejSala) [reply src/lib/client.spec.ts]: I only tested logic contained in this file and used mocks for external functions. 
This particular error throw which you mentioned is in ./api/file remove method.

File: src/lib/api/security.spec.ts (hunk 10 above)
Comment (AndrzejSala) [reply src/lib/api/security.spec.ts]: Fixed

**Pull request merged**